### ISSUE TITLE
SDK manpages bad-whatis-entry

### ISSUE THREAD

**Issue by TheJayMann** [dotnet/sdk#25709]
### Describe the bug
Manpages for multi-level commands for `dotnet` violate Debian's lintian rule for use by `whatis` and `apropos`.

### To Reproduce
I'm not sure how to reproduce this directly, but I do know attempting to build 6.0.102 from https://github.com/dotnet/installer making use of `debmake` and `sbuild`, at the end of a successful build, every man, with exception to `dotnet-vstest.1`, and `dotnet.1` give a lintian warning of `bad-whatis-entry`, as the name for the multi-level commands are space separated, while it is suggested they be separated instead by an underscore, so they can be properly discovered by `whatis` and `apropos`.  As an example, if I am understanding this correctly, for `dotnet-add-package.1`, instead of the line `\f[C]dotnet add package\f[R]...`, it should be `\f[C]dotnet_add_package\f[R]...`.

### Further technical details
https://lintian.debian.org/tags/bad-whatis-entry

**Comment by baronfel** [comment]
I believe these are all generated via a pandoc script that was contributed to the repo. @omajid is this something you can look into? would it be a matter of extending the python scrubber script? 

@TheJayMann is there an easy way to run `debmake/sbuild` so that we could prevent regressions in this area? or is it something that we'd have to leave to the `dotnet/installer` level?

**Comment by TheJayMann** [comment]
The warning should be easily suppressed using a lintian-override, and also, with quite a bit more effort, can actually modify the man pages at the build level by making use of the patches folder.  I just wanted to raise this issue to make sure that at least later releases could have the man pages correct, assuming this doesn't also cause issues with how other distributions want to have man pages defined.

**Comment by omajid** [comment]
I opened a PR at https://github.com/dotnet/sdk/pull/25755 to update the docs in sdk.

I don't have a way to run run lintian to see if this actually fixes everything. @TheJayMann can you run lintian against this?

**Comment by omajid** [comment]
Oh, I also found that https://manpages.debian.org/bullseye/manpages/man.7.en.html says:

> The only mandatory heading is NAME, which should be the first section and be followed on the next line by a one-line description of the program:
>
>     .SH NAME
>     item \- description

I can't find a way to generate the `\-`. Not sure if it will still get flagged or not. A number of other man pages I found have a `-` and that seems to be fine....

**Comment by TheJayMann** [comment]
I looked into the documentation for lintian, and I believe I have a way to check just for this issue.  Rather than having the full lintian process run after sbuild completes, I can run lintian only on a single binary package, and, if necessary, can also limit what checks it is performing.  After this full build that I'm doing completes, I'll try to get the changes from the pull request put on top of my current source build, do a build without lintian, then do a lintian only on the `dotnet-sdk-6.0` package, and, if it takes too long, I'll stop it and only have it do the `bad-whatis-entry` check. This may still take a day, as I still have a previous build being processed to see if I was able to eliminate other lintian issues I had.

**Comment by omajid** [comment]
Thanks, let me dig around some more.

>  I should be able to create a quick source package that will do that if it will help with testing and development.

If you can share it, that would be great!

**Comment by TheJayMann** [comment]
Here is the source package, made up of three files.  The two `.zip` files are not actually zip files, I just added the extension to it so that it could be uploaded.  The `dotnet-sdk-manpages_1.0.0.orig.tar.gz` file contains the manpages from source build 6.0.102 with the patch from the PR applied.  Ideally, you would replace that with a file with the same name and same internal directory layout with updated files.  I was able to run `sbuild` with only these three files by running `sbuild dotnet-sdk-manpages_1.0.0-1.dsc -d unstable`, while already having an sbuild schroot of Debian unstable already set up.  Alternatively, you can extract the contents of `dotnet-sdk-manpages_1.0.0.orig.tar.gz` and `dotnet-sdk-manpages_1.0.0-1.debian.tar.xz` into the same directory and just run `sbuild` from that directory.  I'm sure you can also use another build system, such as `debuild`, `pbuild`, or `dpkg-buildpackage`.  If lintian is not automatically run after the build, you can manually run lintian specifying the produced deb package.
[dotnet-sdk-manpages_1.0.0.orig.tar.gz](https://github.com/dotnet/sdk/files/8827605/dotnet-sdk-manpages_1.0.0.orig.tar.gz)
[dotnet-sdk-manpages_1.0.0-1.debian.tar.xz.zip](https://github.com/dotnet/sdk/files/8827609/dotnet-sdk-manpages_1.0.0-1.debian.tar.xz.zip)
[dotnet-sdk-manpages_1.0.0-1.dsc.zip](https://github.com/dotnet/sdk/files/8827610/dotnet-sdk-manpages_1.0.0-1.dsc.zip)

**Comment by omajid** [comment]
I updated my PR to address some of the issues.

> As far as the remaining bad-whatis-entry warnings, four entries remain, and, from what I can tell, these are the issues. 

Thanks for the great analysis!

> First, with dotnet-environment-variables and dotnet-new-sdk-templates, there does not appear to be a name section. Also, these do not appear to be user commands, thus, would be more appropriate to put then in section 7 rather than 1

Both of those should be fixed now. There's a (generated) name section now, and files named `dotnet-environment-variables.7` and `dotnet-new-sdk-templates.7` are generated.

> The issue with dotnet-install-scripts is ...

The SDK doesnt provide a dotnet-install-script command, so the man page shouldn't have been generated to begin with. It's now removed and not-generated.

>  Finally, with dotnet-new-update, the - between the commands and descriptions is missing

I fixed it in the docs repo (https://github.com/dotnet/docs/pull/29713) and the re-generated man pages look okay to me.

> Also, one oddity I noticed with choosing to use - to separate subcommands, using dotnet-new-list as an example, as the command is dotnet new --list, the name appears as dotnet-new---list. I don't consider this a problem, but I raise it just in case anyone else does.

I am kind of stumped here. Most other commands don't have man pages for individual options. The options are documented in the main man page of the command. I have left this as-is for now.

> Alternatively, you can extract the contents of dotnet-sdk-manpages_1.0.0.orig.tar.gz and dotnet-sdk-manpages_1.0.0-1.debian.tar.xz into the same directory and just run sbuild from that directory

As a complete Debian noob, I tried this out. I don't know what to do about this error, though:

```
omajid@debian:~/work$ sbuild                                                                                      
dh clean                                                                                                                              
   dh_clean                                                                                                                           
dpkg-source: info: using source format '3.0 (quilt)'                                                                                  
dpkg-source: info: building dotnet-sdk-manpages using existing ./dotnet-sdk-manpages_1.0.0.orig.tar.gz                   
dpkg-source: warning: ignoring deletion of directory manpages                                                                         
dpkg-source: warning: ignoring deletion of directory manpages/tool                                                                    
dpkg-source: warning: ignoring deletion of file manpages/tool/update-man-pages.sh, use --include-removal to override
...
dpkg-source: warning: executable mode 0755 of 'dotnet-sdk-manpages-1.0.0/manpages/tool/remove-metadata-and-embed-includes.py' will not be represented in diff                                                                                                               
dpkg-source: warning: executable mode 0755 of 'dotnet-sdk-manpages-1.0.0/manpages/tool/run_docker.sh' will not be represented in diff 
dpkg-source: warning: executable mode 0755 of 'dotnet-sdk-manpages-1.0.0/manpages/tool/update-man-pages.sh' will not be represented in diff                                                                                                                                 
dpkg-source: info: local changes detected, the modified files are:                                                                    
 work/dotnet-sdk-manpages-1.0.0/manpages/sdk/dotnet-add-package.1                                                                     
 work/dotnet-sdk-manpages-1.0.0/manpages/sdk/dotnet-add-reference.1                                                                   
 work/dotnet-sdk-manpages-1.0.0/manpages/sdk/dotnet-build-server.1                                                                    
 work/dotnet-sdk-manpages-1.0.0/manpages/sdk/dotnet-build.1                                                                           
...
 work/dotnet-sdk-manpages-1.0.0/manpages/tool/run_docker.sh                                                                           
 work/dotnet-sdk-manpages-1.0.0/manpages/tool/update-man-pages.sh
dpkg-source: info: you can integrate the local changes with dpkg-source --commit
dpkg-source: error: aborting due to unexpected upstream changes, see /tmp/dotnet-sdk-manpages_1.0.0-1.diff.DOm0gK
E: Failed to package source directory /home/omajid/work
```

Any pointers on how to get past these?

**Comment by TheJayMann** [comment]
I'll admit I'm a bit of a beginner to debian packages as well.  However, my understanding is that, really, when working on debian packages, only the `debian` directory is freely editable.  Any editing you do to upstream sources really should have a patch generated and placed in the `debian/patches` folder.  Essentially, the error here, as far as my knowledge goes is that the source directory does not match the upstream archive after applying patches to the upstream archive.  Really, the simplest way to run these tests would be to simply repack the files into a new upstream archive file, following the same directory structure as the original, and use that to build again.

**Comment by omajid** [comment]
I haven't had any luck with getting a debian lintian run to work. If anyone knows of a guide for total n00bs that I could follow step by step, that would be great.

Otherwise, @TheJayMann can you test out my latest changes in the PR?

**Comment by TheJayMann** [comment]
I can do that in about an hour when I get home.

Also, I think the easiest way to be able to try this out is to delete the `dotnet-sdk-manpages_1.0.0.orig.tar.gz` that I provided, create a new archive with the same name (all of the man pages within the archive should be in the directory `./dotnet-sdk-manpages-1.0.0/manpages/sdk/`, the rest of the files don't matter), then run `sbuild dotnet-sdk-manpages_1.0.0-1.dsc -d unstable`.  Essentially, what I did to create the archive was create a directory `dotnet-sdk-manpages-1.0.0`, copied the `manpages` directory from sdk repository, archived `dotnet-sdk-manpages-1.0.0`, then `debmake` renamed the archive file to `dotnet-sdk-manpages_1.0.0.orig.tar.gz` (technically it created a symlink, but the tools don't know or care about that detail).

**Comment by TheJayMann** [comment]
So, after a little bit more research, I believe setting this up as a native Debian package will make iterative development easier.  Non-native packages essentially have a separate upstream source code from debian package code, where the upstream source code should not be changed, but, rather, any changes necessary should be set up as a set of patch files, and any updates from upstream should be a new version of the package all together.  A native debian package has the debian package code as part of the original source code.  With this set up, the code itself can be modified as often as necessary, no patch files are necessary.  I just now tested it out creating a native debian package for the manpages, I was able to build it, then modify a man page, build it again, and see that the modification was present.  This should make testing this a bit easier.

[dotnet-sdk-manpages_1.0.0.tar.xz.zip](https://github.com/dotnet/sdk/files/8846837/dotnet-sdk-manpages_1.0.0.tar.xz.zip)

This should be as simple as extract the contents, switch to the `dotnet-sdk-manpages-1.0.0` directory, then run `sbuild`, assuming you have properly set up a proper unstable schroot for sbuild.
